File: src/BusinessLogic/Handlers/MessageHandlers/WatchMessageHandler.cs
using Bounan.Bot.BusinessLogic.Clients.Shikimori;
using Bounan.Bot.BusinessLogic.Clients.Shikimori.Models;
using Bounan.Bot.BusinessLogic.CommandDto;
using Bounan.Bot.BusinessLogic.Configs;
using Bounan.Bot.BusinessLogic.Helpers;
using Bounan.Bot.BusinessLogic.Interfaces;
using Bounan.Bot.BusinessLogic.Models;
using Bounan.Bot.BusinessLogic.Services;
using Bounan.Bot.TelegramBot.Telegram;
using Bounan.Common.Enums;
using Bounan.Common.Models;
using Bounan.LoanApi.Interfaces;
using Bounan.LoanApi.Models;
using Microsoft.Extensions.Logging;
using Microsoft.Extensions.Options;
using Telegram.Bot;
using Telegram.Bot.Types;
using Telegram.Bot.Types.Enums;
using Telegram.Bot.Types.ReplyMarkups;

namespace Bounan.Bot.BusinessLogic.Handlers.MessageHandlers;

public class WatchMessageHandler : IMessageHandler
{
    private readonly TelegramBotConfig _telegramBotConfig;

    public WatchMessageHandler(
        ILogger<WatchMessageHandler> logger,
        IShikimoriApi shikimoriApi,
        ILoanApiComClient botLoanApiComClient,
        IAniManClient aniManClient,
        ITelegramBotClient botClient,
        IOptions<TelegramBotConfig> telegramBotConfig,
        IFileIdFinder fileIdFinder)
    {
        _telegramBotConfig = telegramBotConfig.Value;
        Logger = logger;
        ShikimoriApi = shikimoriApi;
        BotLoanApiComClient = botLoanApiComClient;
        AniManClient = aniManClient;
        BotClient = botClient;
        FileIdFinder = fileIdFinder;
    }

    private ILogger Logger { get; }

    private IShikimoriApi ShikimoriApi { get; }

    private ILoanApiComClient BotLoanApiComClient { get; }

    private IAniManClient AniManClient { get; }

    private ITelegramBotClient BotClient { get; }

    private IFileIdFinder FileIdFinder { get; }

    public static bool CanHandle(Message message) => message.Text?.StartsWith(WatchCommandDto.Command) ?? false;

    public async Task HandleAsync(Message message, CancellationToken cancellationToken)
    {
        ArgumentNullException.ThrowIfNull(message);
        ArgumentNullException.ThrowIfNull(message.Text);
        ArgumentNullException.ThrowIfNull(message.Chat.Id);

        var commandDto = CommandConvert.DeserializeCommand<WatchCommandDto>(message.Text);
        if (commandDto is null)
        {
            Logger.LogInformation("Failed to deserialize command {Command}", message.Text);
            return;
        }

        var searchResults = await BotLoanApiComClient.GetExistingVideos(commandDto.MyAnimeListId, cancellationToken);
        if (searchResults is null or { Count: 0 })
        {
            Logger.LogWarning("No videos for {MyAnimeListId}", commandDto.MyAnimeListId);
            await BotClient.SendTextMessageAsync(
                message.Chat.Id,
                "Этого аниме (пока?) нет в базе",
                cancellationToken: cancellationToken);
            return;
        }

        var episodesInDub = searchResults
            .Where(item => AnimeHelpers.DubToKey(item.Dub) == AnimeHelpers.DubToKey(commandDto.Dub))
            .Select(item => item.Episode)
            .OrderBy(ep => ep)
            .ToArray();
        ArgumentOutOfRangeException.ThrowIfZero(episodesInDub.Length);

        var selectedVideo = searchResults
            .FirstOrDefault(v => AnimeHelpers.DubToKey(v.Dub) == AnimeHelpers.DubToKey(commandDto.Dub)
                                 && v.Episode == commandDto.Episode);
        if (selectedVideo is null)
        {
            Logger.LogInformation("Video not found in dub");
            await SendSwitchDubButtonsAsync(message.Chat.Id, searchResults, commandDto.Episode);
            return;
        }

        var botRequest = new BotRequest(
            MyAnimeListId: selectedVideo.MyAnimeListId,
            Dub: selectedVideo.Dub,
            Episode: selectedVideo.Episode,
            ChatId: message.Chat.Id);
        var videoInfo = await AniManClient.GetAnimeAsync(botRequest, CancellationToken.None);

        var keyboard = TelegramHelpers.GetKeyboard(
            commandDto,
            searchResults.Select(v => v.Episode),
            _telegramBotConfig.ButtonsPagination);

        switch (videoInfo?.Status)
        {
            case VideoStatus.Pending:
            case VideoStatus.Downloading:
                Logger.LogInformation("Video not downloaded");
                await BotClient.SendTextMessageAsync(
                    message.Chat.Id,
                    "Видео готовится. Я пришлю его, как только будет готово",
                    replyMarkup: keyboard,
                    cancellationToken: cancellationToken);
                break;

            case VideoStatus.Failed:
            case VideoStatus.NotAvailable:
                Logger.LogInformation("Video failed to download");
                await BotClient.SendTextMessageAsync(
                    message.Chat.Id,
                    "Я не смог найти эту серию. Какая-то ошибка, разработчик уже уведомлен",
                    replyMarkup: keyboard,
                    cancellationToken: cancellationToken);
                break;

            case VideoStatus.Downloaded:
                await SendVideoAsync(message, commandDto, videoInfo, keyboard, cancellationToken);
                break;

            case VideoStatus.Unknown:
            case null:
                Logger.LogError("Lambda returned null");
                await BotClient.SendTextMessageAsync(
                    message.Chat.Id,
                    "Что-то сломалось. Разработчик уже уведомлен",
                    replyMarkup: keyboard,
                    cancellationToken: cancellationToken);
                break;

            default:
                throw new InvalidOperationException($"Unknown video status: {videoInfo.Status}");
        }
    }

    private async Task SendVideoAsync(
        Message message,
        IVideoKey commandDto,
        IBotResponse videoInfo,
        IReplyMarkup keyboard,
        CancellationToken cancellationToken)
    {
        ArgumentNullException.ThrowIfNull(videoInfo.MessageId);

        AnimeInfo? animeInfo;
        try
        {
            animeInfo = await ShikimoriApi.GetAnimeInfoAsync(commandDto.MyAnimeListId, cancellationToken);
            Logger.LogInformation(
                "Got anime info for {MyAnimeListId}: {AnimeInfo}",
                commandDto.MyAnimeListId,
                animeInfo);
        }
        catch (Exception ex)
        {
            Logger.LogError(ex, "Failed to get anime info");
            animeInfo = null;
        }

        var fileId = await FileIdFinder.GetFileIdAsync(videoInfo.MessageId.Value);
        ArgumentNullException.ThrowIfNull(fileId);

        await BotClient.SendVideoAsync(
            message.Chat.Id,
            new InputFileId(fileId),
            caption: TelegramHelpers.GetVideoDescription(animeInfo, commandDto),
            parseMode: ParseMode.Html,
            replyMarkup: keyboard,
            cancellationToken: cancellationToken);
    }

    private async Task SendSwitchDubButtonsAsync(long chatId, IEnumerable<IVideoKeyWithLink> searchResults, int episode)
    {
        var inOtherDubs = searchResults
            .Where(item => item.Episode == episode)
            .OrderBy(item => item.Dub)
            .ToArray();

        Logger.LogWarning("Episode not found in dub. Other dubs: {Dubs}", inOtherDubs.Select(ep => ep.Dub));

        await BotClient.SendTextMessageAsync(
            chatId,
            "Серии нет в этом переводе. Попробуйте другой",
            replyMarkup: new InlineKeyboardMarkup(
                inOtherDubs.Select(ep => new[]
                {
                    new InlineKeyboardButton(ep.Dub)
                    {
                        CallbackData = CommandConvert.SerializeCommand(
                            new WatchCommandDto
                            {
                                MyAnimeListId = ep.MyAnimeListId,
                                Dub = AnimeHelpers.DubToKey(ep.Dub),
                                Episode = episode,
                            }),
                    },
                })));
    }
}

File: src/handlers/on-webhook/event-handlers/message-handlers/watch-message-handler.ts
import { Message } from '@lightweight-clients/telegram-bot-api-lightweight-client';
import {
    copyMessage,
    CopyMessageData,
    InlineKeyboardMarkup,
    sendMessage,
} from '@lightweight-clients/telegram-bot-api-lightweight-client';

import { getVideoInfo } from '../../../../api-clients/animan/animan-client';
import { getAllExistingVideos } from '../../../../api-clients/cached-loan-api-client';
import { getShikiAnimeInfo } from '../../../../api-clients/cached-shikimori-client';
import { config } from '../../../../config/config';
import { assert } from '../../../../shared/helpers/assert';
import { dubToKey } from '../../../../shared/helpers/dub-to-key';
import { BotResponse, VideoKey } from '../../../../shared/models';
import { getKeyboard } from '../../../../shared/telegram/get-keyboard';
import { getVideoDescription } from '../../../../shared/telegram/get-video-description';
import { Texts } from '../../../../shared/telegram/texts';
import { subscribeOneTime } from '../../../subscriptions-repository';
import { WatchCommandDto } from '../../command-dtos';
import { MessageHandler } from '../query-handler';

const sendSwitchDubButtons = async (chatId: number, searchResults: VideoKey[], episode: number) => {
    const inOtherDubs = searchResults
        .filter(item => item.episode === episode)
        .sort((a, b) => a.dub.localeCompare(b.dub));

    console.log('Episode not found in dub. Other dubs: ', inOtherDubs.map(ep => ep.dub));

    await sendMessage({
        chat_id: chatId,
        text: Texts.Message__EpisodeWithDubNotFound,
        reply_markup: {
            inline_keyboard: inOtherDubs.map(ep => [{
                text: ep.dub,
                callback_data: new WatchCommandDto(ep.myAnimeListId, dubToKey(ep.dub), episode).toString(),
            }]),
        },
    });
}

const sendVideo = async (
    message: Pick<Message, 'chat' | 'text'>,
    commandDto: WatchCommandDto,
    videoInfo: BotResponse,
    keyboard: InlineKeyboardMarkup,
) => {
    const animeInfo = await getShikiAnimeInfo(commandDto.myAnimeListId);
    const videoDescription = getVideoDescription(animeInfo, commandDto, videoInfo.scenes);

    const args: CopyMessageData = {
        chat_id: message.chat.id,
        from_chat_id: config.value.telegram.videoChatId,
        message_id: videoInfo.messageId!,
        caption: videoDescription,
        reply_markup: keyboard,
        parse_mode: 'HTML',
    };

    console.log('Sending video: ', JSON.stringify(args));
    await copyMessage(args);
}

const sendVideoResult = async (
    message: Pick<Message, 'chat' | 'text'>,
    commandDto: WatchCommandDto,
    searchResults: VideoKey[],
    selectedVideo: VideoKey,
) => {
    const episodesInDub = searchResults
        .filter(item => dubToKey(item.dub) === commandDto.dub)
        .map(item => item.episode)
        .sort((a, b) => a - b);
    if (episodesInDub.length === 0) {
        throw new Error('No episodes found');
    }

    const videoInfo = await getVideoInfo(selectedVideo);
    console.log('Anime info: ', JSON.stringify(videoInfo));

    const keyboard = getKeyboard(commandDto, episodesInDub, videoInfo?.publishingDetails);

    switch (videoInfo?.status) {
        case 'Pending':
        case 'Downloading':
            console.log('Video not downloaded');
            await Promise.all([
                await subscribeOneTime(selectedVideo, message.chat.id),
                await sendMessage({
                    chat_id: message.chat.id,
                    text: Texts.Message__VideoIsCooking,
                    reply_markup: keyboard,
                }),
            ]);
            break;

        case 'Failed':
        case 'NotAvailable':
            console.log('Video failed to download');
            await sendMessage({
                chat_id: message.chat.id,
                text: Texts.ErrorOnEpisode,
                reply_markup: keyboard,
            });
            break;

        case 'Downloaded':
            console.log('Sending video');
            await sendVideo(message, commandDto, videoInfo, keyboard);
            break;

        case null:
            console.error('Lambda returned null');
            await sendMessage({
                chat_id: message.chat.id,
                text: Texts.UnknownError,
                reply_markup: keyboard,
            });
            break;

        default:
            throw new Error(`Unknown video status: ${videoInfo.status}`);
    }
}

const canHandle = (message: Message): boolean => message.text?.startsWith(WatchCommandDto.Command) ?? false;

const handler: MessageHandler = async (message) => {
    assert(!!message.text);
    assert(!!message.chat?.id);

    console.log('Received watch command');

    const commandDto = WatchCommandDto.fromPayload(message.text!) as WatchCommandDto;
    console.log('Parsed command: ', commandDto);
    if (!commandDto) {
        console.warn('Failed to deserialize command', message.text);
        return;
    }

    const searchResults = await getAllExistingVideos(commandDto.myAnimeListId);
    if (!searchResults || searchResults.length === 0) {
        console.log('No videos found');
        await sendMessage({
            chat_id: message.chat.id,
            text: Texts.Search__NoResultsInLoan,
        });
        return;
    }

    const selectedVideo = searchResults
        .find(v => dubToKey(v.dub) === commandDto.dub && v.episode === commandDto.episode);
    console.log('Selected video: ', selectedVideo);
    if (!selectedVideo) {
        console.warn('Video not found in dub');
        await sendSwitchDubButtons(message.chat.id, searchResults, commandDto.episode);
        return;
    }

    await sendVideoResult(message, commandDto, searchResults, selectedVideo);

    console.log('Watch command handled');
}

export const watchMessageHandler = {
    canHandle,
    handler,
};